Make Sage work with maxima 5.47

antonio-rojas opened the pull request:
Maxima 5.47 has been released. The main change that affects Sage is https://sourceforge.net/p/maxima/code/ci/f5ca00582c24cfedca53a664335f6c13e1e40cf9, which makes the maxima ecl library crash because the `e-val` variable is unassigned. We fix this by running the `initialize-runtime-globals` function on load.

The remaining issues are a few harmless test failures caused by numerical noise and output format changes (such as additional parentheses near minus signs)

antonio-rojas committed:
Run initialize-runtime-globals on maxima initialization to fix crashes with maxima 5.47

antonio-rojas committed:
Make tests pass with maxima 5.47

antonio-rojas committed:
Fix one more test with maxima 5.47

tornaria commented on src/doc/de/tutorial/interfaces.rst:
```diff
@@ -273,7 +273,7 @@ deren :math:`i,j` Eintrag gerade :math:`i/j` ist, für :math:`i,j=1,\ldots,4`.
     sage: A.eigenvalues()
     [[0,4],[3,1]]
     sage: A.eigenvectors()
```
```suggestion
    sage: A.eigenvectors().sage()
```
seems to produce more stable result, what do you think?

tornaria commented on src/doc/de/tutorial/interfaces.rst:
```diff
@@ -335,7 +335,7 @@ Und der letzte ist die berühmte Kleinsche Flasche:
     sage: maxima("expr_1: 5*cos(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0) - 10.0")
     5*cos(x)*(sin(x/2)*sin(2*y)+cos(x/2)*cos(y)+3.0)-10.0
     sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)")
```
```suggestion
    sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)").sage()
```

tornaria commented on src/doc/de/tutorial/interfaces.rst:
```diff
@@ -273,7 +273,7 @@ deren :math:`i,j` Eintrag gerade :math:`i/j` ist, für :math:`i,j=1,\ldots,4`.
     [[0,4],[3,1]]
     sage: A.eigenvectors()
-    [[[0,4],[3,1]],[[[1,0,0,-4],[0,1,0,-2],[0,0,1,-4/3]],[[1,2,3,4]]]]
+    [[[0,4],[3,1]],[[[1,0,0,-4],[0,1,0,-2],[0,0,1,-...4/3...]],[[1,2,3,4]]]]
```
```suggestion
    [[[0, 4], [3, 1]],
     [[[1, 0, 0, -4], [0, 1, 0, -2], [0, 0, 1, -4/3]], [[1, 2, 3, 4]]]]
```

tornaria commented on src/doc/de/tutorial/interfaces.rst:
```diff
@@ -335,7 +335,7 @@ Und der letzte ist die berühmte Kleinsche Flasche:
     5*cos(x)*(sin(x/2)*sin(2*y)+cos(x/2)*cos(y)+3.0)-10.0
     sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)")
-    -5*sin(x)*(sin(x/2)*sin(2*y)+cos(x/2)*cos(y)+3.0)
+    -...5*sin(x)*(sin(x/2)*sin(2*y)+cos(x/2)*cos(y)+3.0)...
```
```suggestion
    -5*(cos(1/2*x)*cos(y) + sin(1/2*x)*sin(2*y) + 3.0)*sin(x)
```
same here

tornaria commented on src/doc/de/tutorial/tour_algebra.rst:
```diff
@@ -211,7 +211,7 @@ Lösung: Berechnen Sie die Laplace-Transformierte der ersten Gleichung
 
     sage: de1 = maxima("2*diff(x(t),t, 2) + 6*x(t) - 2*y(t)")
     sage: lde1 = de1.laplace("t","s"); lde1
```
```suggestion
    sage: lde1 = de1.laplace("t","s"); lde1.sage()
```
Here as well, the output seems easier to read.

tornaria commented on src/doc/de/tutorial/tour_algebra.rst:
```diff
@@ -211,7 +211,7 @@ Lösung: Berechnen Sie die Laplace-Transformierte der ersten Gleichung
     sage: de1 = maxima("2*diff(x(t),t, 2) + 6*x(t) - 2*y(t)")
     sage: lde1 = de1.laplace("t","s"); lde1
-    2*((-%at('diff(x(t),t,1),t = 0))+s^2*'laplace(x(t),t,s)-x(0)*s) -2*'laplace(y(t),t,s)+6*'laplace(x(t),t,s)
+    2*(...-...%at('diff(x(t),t,1),t = 0))+s^2*'laplace(x(t),t,s)-x(0)*s) -2*'laplace(y(t),t,s)+6*'laplace(x(t),t,s)
```
```suggestion
    2*s^2*laplace(x(t), t, s) - 2*s*x(0) + 6*laplace(x(t), t, s) - 2*laplace(y(t), t, s) - 2*D[0](x)(0)
```

tornaria commented on src/sage/calculus/calculus.py:
```diff
@@ -1336,7 +1336,7 @@ def limit(ex, dir=None, taylor=False, algorithm='maxima', **argv):
         0
         sage: limit(floor(x), x=0)
-        und
+        ...nd
```
What happened here? Answer was `und` and now is `ind`, is there a semantic difference? Who is choosing the naming?

antonio-rojas replied:
See https://sourceforge.net/p/maxima/code/ci/df1748a88b2906aac866297a2cec48747e2d3728

tornaria commented on src/sage/functions/bessel.py:
```diff
@@ -295,7 +295,7 @@ class Function_Bessel_J(BuiltinFunction):
         sage: m = maxima(bessel_J(2, x))
         sage: m.integrate(x)
-        (hypergeometric([3/2],[5/2,3],-_SAGE_VAR_x^2/4)*_SAGE_VAR_x^3)/24
+        (hypergeometric([3/2],[5/2,3],-..._SAGE_VAR_x^2/4)...*_SAGE_VAR_x^3)/24
```
Much nicer IMO:
```
sage: m = maxima(bessel_J(2, x))
sage: m.integrate(x).sage()
1/24*x^3*hypergeometric((3/2,), (5/2, 3), -1/4*x^2)
```

antonio-rojas committed:
Nicer fixes for tests

tornaria commented on src/doc/en/constructions/linear_algebra.rst:
```diff
@@ -278,7 +278,7 @@ Another approach is to use the interface with Maxima:
     sage: eig = A.eigenvectors()
     sage: eig
-    [[[-sqrt(3)*%i,sqrt(3)*%i],[1,1]], [[[1,(sqrt(3)*%i+1)/4]],[[1,-(sqrt(3)*%i-1)/4]]]]
+    [[[-...sqrt(3)*%i...,sqrt(3)*%i],[1,1]], [[[1,(sqrt(3)*%i+1)/4]],[[1,-...(sqrt(3)*%i-1)/4...]]]]
```
missed this  one

orlitzky replied:
I cannot imagine why we would tell the poor user to use the low-level maxima interface to find eigenvectors in a linear algebra tutorial. My opinion: delete the bad advice (and the failing test).

tornaria commented on src/sage/functions/orthogonal_polys.py:
```diff
@@ -974,7 +974,7 @@ def __init__(self):
             chebyshevu(x, x)
             sage: maxima(chebyshev_U(2,x, hold=True))
-            3*((-(8*(1-_SAGE_VAR_x))/3)+(4*(1-_SAGE_VAR_x)^2)/3+1)
+            3*(...-...(8*(1-_SAGE_VAR_x))/3)+(4*(1-_SAGE_VAR_x)^2)/3+1)
```
```
sage: maxima(chebyshev_U(2,x, hold=True)).sage()
4*(x - 1)^2 + 8*x - 5
```
I don't know what this is trying to acomplish... I guess that maxima knows how to handle  `chebyshev_U(2, x)`.

antonio-rojas replied:
I guess it's testing conversion to maxima, so I'm not sure converting back to sage the result makes much sense

orlitzky replied:
Yeah, it's supposed to be an example of the `__init__` method. Why send it to maxima at all?

tornaria commented on src/sage/functions/special.py:
```diff
@@ -457,7 +457,7 @@ class EllipticE(BuiltinFunction):
         sage: # this is still wrong: must be abs(sin(z)) + 2*round(z/pi)
         sage: elliptic_e(z, 1).simplify()
-        2*round(z/pi) + sin(z)
+        2*round(z/pi) ... sin(...z)
```
Really, it seems to me the previous answer is wrong, and the new answer is correct, even if it's not the nicest (note that `abs(sin(z)) + 2*round(z/pi)` is also incorrect). Let's just write something like
```
sage: elliptic_e(z, 1).simplify()                # not tested: bad answer with maxima < 5.47
2*round(z/pi) - sin(pi*round(z/pi) - z)
```
(is there a way to flag a test as "expected to fail" rather than "not test" ?)

This is an even better test which shows the new answer is ok:
```
sage: f(z) = elliptic_e(z, 1).simplify()
sage: all([abs(f(z).simplify()-elliptic_e(z,1))<1e-15
....:      for z in [RR.random_element(10) for _ in range(10)]])
True
```
you can see this passes with maxima 5.47 but fails with maxima 5.46.

tornaria replied:
Could do something like:
```
sage: maxima.version() < '5.47' or all([
....:     abs(f(z).simplify()-elliptic_e(z,1))<1e-15
....:     for z in [RR.random_element(10) for _ in range(100)]])
True
```
but it's a bit ugly :shrug:   (ideal would be a way to label the test something  like `xfail when maxima.version() < '5.47'`.

tornaria commented on src/sage/interfaces/interface.py:
```diff
@@ -1581,18 +1581,18 @@ def _mul_(self, right):
+            -...cos(x)*sin(x)...
             sage: _(2)
-            -cos(2)*sin(2)
+            -...cos(2)*sin(2)...
```
I give up... :fearful:  At some point these tests are just not useful anymore... with all this `...` stuffed in.

orlitzky replied:
It's still not a great solution, but if we drop the negative sign from `'-cos(x)`, then the answer is consistent.

orlitzky commented on src/doc/en/tutorial/interfaces.rst:
```diff
@@ -320,8 +320,8 @@ The next plot is the famous Klein bottle (do not type the ``....:``)::
-    sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)")
-    -5*sin(x)*(sin(x/2)*sin(2*y)+cos(x/2)*cos(y)+3.0)
+    sage: maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)").sage()
+    -5*(cos(1/2*x)*cos(y) + sin(1/2*x)*sin(2*y) + 3.0)*sin(x)
```
These don't actually need to be echoed for the example. I'd suggest e.g.

```
sage: _ = maxima("expr_2: -5*sin(x)*(cos(x/2)*cos(y) + sin(x/2)*sin(2*y)+ 3.0)")
```

instead.

antonio-rojas committed:
Improve one more test output

antonio-rojas committed:
Disable test that gives wrong answer with maxima < 5.47

orlitzky commented on src/sage/functions/bessel.py:
```diff
@@ -1121,8 +1121,8 @@ def Bessel(*args, **kwds):
-        sage: f.derivative('_SAGE_VAR_y')
-        -(bessel_k(_SAGE_VAR_x+1,_SAGE_VAR_y)+bessel_k(_SAGE_VAR_x-1, _SAGE_VAR_y))/2
+        sage: f.derivative('_SAGE_VAR_y').sage()
+        -1/2*bessel_K(x + 1, y) - 1/2*bessel_K(x - 1, y)
```
I don't think we really care about the exact string output here; we want to know that going to maxima, taking a derivative, and going back to sage gives you the same result that taking the derivative in sage would:

```
sage: x,y = SR.var('x,y')
sage: f = Bessel(typ='K')(x,y)
sage: expected = f.derivative(y)
sage: actual = maxima(f).derivative('_SAGE_VAR_y').sage()
sage: bool(actual == expected)
True
```

orlitzky commented on src/sage/functions/other.py:
```diff
@@ -501,7 +501,7 @@ def __init__(self):
             floor(x + 5.40000000000000)
             sage: a.simplify()
-            floor(x + 0.4000000000000004) + 5
+            floor(x + 0.400000000000000...) + 5
```
This looks like it works without the fuzz (because 1/4 is a power of two):

```
sage: a = floor(5.25 + x); a
floor(x + 5.25000000000000)
sage: a.simplify()
floor(x + 0.25) + 5
```

but the best example is probably,

```
sage: a = floor(27/5 + x); a
floor(x + 27/5)
sage: a.simplify()
floor(x + 2/5) + 5
```

orlitzky commented on src/sage/functions/special.py:
```diff
@@ -455,9 +455,8 @@ class EllipticE(BuiltinFunction):
-        sage: # this is still wrong: must be abs(sin(z)) + 2*round(z/pi)
-        sage: elliptic_e(z, 1).simplify()
-        2*round(z/pi) + sin(z)
+        sage: elliptic_e(z, 1).simplify() # not tested - gives wrong answer with maxima < 5.47
```
I guess this is OK for now. This is testing for a specific maxima bug (3967) that was fixed in 5.47, and that is already part of the maxima test suite. So normally, I would say "delete it," but there aren't many other examples for `elliptic_e` and showing that it can be simplified to a more-explicit form is nice.

orlitzky commented on src/sage/interfaces/maxima.py:
```diff
@@ -51,7 +51,7 @@
     sage: F
-    -(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)
+    -...(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)...
     sage: type(F)
```
This is more resilient as e.g.

```
sage: x,y = SR.var('x,y')
sage: actual = maxima.factor('x^5 - y^5').sage()
sage: expected = -(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)
sage: bool(actual == expected)
```

orlitzky commented on src/sage/interfaces/maxima.py:
```diff
@@ -72,17 +72,17 @@
     sage: repr(F)
-    '-(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)'
+    '-...(y-x)*(y^4+x*y^3+x^2*y^2+x^3*y+x^4)...'
     sage: F.str()
```
There's no reason to continue using the (complicated) expression `F` here. Something predictable like `maxime(1/2)` should work fine.

orlitzky commented on src/sage/interfaces/maxima_abstract.py:
```diff
@@ -1573,7 +1573,7 @@ def integral(self, var='x', min=None, max=None):
 
             sage: f = maxima('exp(x^2)').integral('x',0,1); f
-            -(sqrt(%pi)*%i*erf(%i))/2
+            -...(sqrt(%pi)*%i*erf(%i))/2...
```
You can call `.sage()` on this one and compare it to the expected answer with `==`.

orlitzky commented on src/sage/modules/free_module_element.pyx:
```diff
@@ -4054,7 +4054,7 @@ cdef class FreeModuleElement(Vector):   # abstract base class
             sage: vec,answers=r.nintegral(t,0,1)
             sage: vec
-            (0.5, 0.3333333333333334, 0.4596976941318602)
+            (0.5, 0.333333333333333..., 0.4596976941318602...)
```
abs tol here?

orlitzky commented on src/sage/symbolic/relation.py:
```diff
@@ -659,7 +659,7 @@ def solve(f, *args, **kwds):
         sage: sols = solve([x^3==y,y^2==x], [x,y]); sols[-1], sols[0]
         ([x == 0, y == 0],
-         [x == (0.3090169943749475 + 0.9510565162951535*I),
+         [x == (0.309016994374947... + 0.9510565162951535*I),
```
abs tol here?

orlitzky commented on src/doc/en/tutorial/interfaces.rst:
```diff
@@ -267,8 +267,8 @@ whose :math:`i,j` entry is :math:`i/j`, for
-    sage: A.eigenvectors()
-    [[[0,4],[3,1]],[[[1,0,0,-4],[0,1,0,-2],[0,0,1,-4/3]],[[1,2,3,4]]]]
+    sage: A.eigenvectors().sage()
+    [[[0, 4], [3, 1]], [[[1, 0, 0, -4], [0, 1, 0, -2], [0, 0, 1, -4/3]], [[1, 2, 3, 4]]]]
```
This is difficult to test correctly because the eigenvalues, eigenvectors, and multiplicities are split across three lists and need to be matched up. The eigenvectors I would not expect to be unique anyway.

Lazy fix: switch it to a matrix with only nonnegative eigenvalues and a set of eigenvectors whose entries can be chosen integral and nonnegative.

antonio-rojas committed:
Modify tests so they give the same answer with maxima 5.46/47

antonio-rojas committed:
Make some tests less verbose

antonio-rojas committed:
Drop no longer needed sage conversion

antonio-rojas committed:
Improve some tests

orlitzky commented on src/sage/functions/bessel.py:
```diff
@@ -294,8 +294,8 @@ class Function_Bessel_J(BuiltinFunction):
         sage: f.integrate(x)
         1/24*x^3*hypergeometric((3/2,), (5/2, 3), -1/4*x^2)
         sage: m = maxima(bessel_J(2, x))
-        sage: m.integrate(x)
```
I think this is a pretty useless test. What does anyone care about the maxima interface here?

In any case, `f.integrate()` _already_ uses maxima by default. We could specify `f.integrate(x,algorithm='maxima')` to ensure that the maxima interface is still tested in case the default ever changes... but again, who wants to read about the maxima interface here?

orlitzky commented on src/doc/en/constructions/linear_algebra.rst:
```diff
@@ -277,8 +277,8 @@ Another approach is to use the interface with Maxima:
     sage: A = maxima("matrix ([1, -4], [1, -1])")
     sage: eig = A.eigenvectors()
-    sage: eig
-    [[[-sqrt(3)*%i,sqrt(3)*%i],[1,1]], [[[1,(sqrt(3)*%i+1)/4]],[[1,-(sqrt(3)*%i-1)/4]]]]
```
I thought I commented on this one already, but don't see it. I don't know why we're telling people to use the low-level maxima interface to compute the eigenvalues of a 2x2 matrix in a linear algebra tutorial. Personally I would just delete it (and the surrounding explanation).

antonio-rojas replied:
Then the gap section immediately after that should be dropped too. I think we should leave any further doc cleanup for a follow up ticket.

orlitzky commented on src/doc/en/tutorial/tour_algebra.rst:
```diff
@@ -217,8 +217,8 @@ the notation :math:`x=x_{1}`, :math:`y=x_{2}`):
-    sage: lde1 = de1.laplace("t","s"); lde1
-    2*((-%at('diff(x(t),t,1),t = 0))+s^2*'laplace(x(t),t,s)-x(0)*s) -2*'laplace(y(t),t,s)+6*'laplace(x(t),t,s)
+    sage: lde1 = de1.laplace("t","s"); lde1.sage()
+    2*s^2*laplace(x(t), t, s) - 2*s*x(0) + 6*laplace(x(t), t, s) - 2*laplace(y(t), t, s) - 2*D[0](x)(0)
```
The next paragraph tries to explain the output from this example, and so no longer applies.

But instead of changing it, this is another place where I don't understand why we're telling people to use the maxima interface. Sage supports laplace transforms itself. The (explicit) round trip through maxima and back can be avoided:

```
sage: t,s = SR.var('t,s')
sage: x = function('x')
sage: y = function('y')
sage: f = 2*x(t).diff(t,2) + 6*x(t) - 2*y(t)
sage: f.laplace(t,s)
2*s^2*laplace(x(t), t, s) - 2*s*x(0) + 6*laplace(x(t), t, s) - 2*laplace(y(t), t, s) - 2*D[0](x)(0)
```

antonio-rojas committed:
Drop redundant test

antonio-rojas committed:
Don't use maxima to compute the Laplace transform in the tutorial

orlitzky reviewed (approved):
Ok, we shouldn't hold this up any longer. The first commit is pretty important. Thanks.

tornaria commented on src/sage/interfaces/maxima_lib.py:
```diff
@@ -133,6 +133,7 @@
     ecl_eval("(require 'maxima \"{}\")".format(MAXIMA_FAS))
 else:
     ecl_eval("(require 'maxima)")
+ecl_eval("(maxima::initialize-runtime-globals)")
```
It seems this is unfixing #26968, as seen here: https://github.com/void-linux/void-packages/actions/runs/5366826557/jobs/9736459766?pr=44624#step:7:43488

Presumably, calling `initialize-runtime-globals` will run `set-pathnames` and fail without the workaround that is in the `try` block... Can this be moved *after* that?

tornaria replied:
My suggestion:

```diff
--- a/src/sage/interfaces/maxima_lib.py
+++ b/src/sage/interfaces/maxima_lib.py
@@ -133,12 +133,11 @@ if MAXIMA_FAS:
     ecl_eval("(require 'maxima \"{}\")".format(MAXIMA_FAS))
 else:
     ecl_eval("(require 'maxima)")
-ecl_eval("(maxima::initialize-runtime-globals)")
 ecl_eval("(in-package :maxima)")
-ecl_eval("(setq $nolabels t))")
-ecl_eval("(defvar *MAXIMA-LANG-SUBDIR* NIL)")
 ecl_eval("(set-locale-subdir)")
 
+# this has to happen early so the workaround works
+# do not add other initialization before this block
 try:
     ecl_eval("(set-pathnames)")
 except RuntimeError:
@@ -155,6 +154,8 @@ except RuntimeError:
     # Call `(set-pathnames)` again to complete its job.
     ecl_eval("(set-pathnames)")
 
+ecl_eval("(initialize-runtime-globals)")
+ecl_eval("(setq $nolabels t))")
 ecl_eval("(defun add-lineinfo (x) x)")
 ecl_eval('(defun principal nil (cond ($noprincipal (diverg)) ((not pcprntd) (merror "Divergent Integral"))))')
 ecl_eval("(remprop 'mfactorial 'grind)")  # don't use ! for factorials (#11539)
```

tornaria committed:
Make sure #26968 is not unfixed after #35707.

Calling initialize-runtime-globals will run set-pathnames and be subject
to the issue described in #26968.

Thus the workaround introduced in #35195 has to be done before anything
that may call set-pathnames (e.g. initialize-runtime-globals).

antonio-rojas committed:
Merge pull request #2 from tornaria/maxima-5.47.0

Make sure #26968 is not unfixed after #35707.
